proposals/20200828-structured-exception-handling.md

```diff
@@ -0,0 +1,214 @@
+
+```
+- rule: Write below binary dir
+  desc: an attempt to write to any file below a set of binary directories
+  condition: >
+    bin_dir and evt.dir = < and open_write
+    and not package_mgmt_procs
+    and not exe_running_docker_save
+    and not python_running_get_pip
+    and not python_running_ms_oms
+    and not user_known_write_below_binary_dir_activities
+  exceptions:
+   - proc_writer: [proc.name, fd.directory]
+   - container_writer: [container.image.repository, fd.directory]
+```
+
+This rule defines two kinds of exeptions: one called proc_writer with a combination of proc.name and fd.directory, and a second called container_writer with a combination of container.image.repository and fd.directory.
+
+Notice that exceptions are defined as a part of the rule. This is important because the author of the rule defines what construes a valid exception to the rule. In this case, an exception can consist of a process and file directory (actor and target), but not a process name only (too broad).
+
+We'll add a new object exception that defines exceptions to a rule:
+
+```
+- exception: Write below binary dir
+  items:
+  - proc_writer:
+    - [apk, /usr/lib/alpine]
+    - [npm, /usr/node/bin]
+  - container_writer:
+    - [docker.io/alpine, /usr/libexec/alpine]
+```
+
+The name of the exception links it to the rule.
+
+A rule exception applies if for a given event, the fields in a rule.exception match all of the values in some exception.item. For example, if a program `apk` writes to a file below `/usr/lib/alpine`, the rule will not trigger, even if the condition is met.
+
+Append will be supported for exception objects. If append: is true, the items in the second definition will be added to the items in the earlier definition. For example, adding:
+
+```
+- exception: Write below binary dir
+  items:
+  - container_writer:
+    - [docker.io/golang-alpine, /usr/libexec/go]
+  append: true
+```
+
+Would add a second container_writer exception.
+
+### Implementation
+
+Each exception can be thought of as an implicit "and not (field1=val1 and field2=val2 and...)" appended to the rule's condition. In practice, that's how exceptions will be implemented.
+
+When a rule is parsed, the original condition will be wrapped in an extra layer of parentheses and all exception values will be appended to the condition. For example, using the example above, the resulting condition will be:
+
+```
+(<Write below binary dir condition>) and not ((proc.name=apk and fd.directory=/usr/lib/alpine) or (proc.name=npm and fd.directory=/usr/node/bin) or (container.image.repository=docker.io/alpine and fd.directory=/usr/libexec/alpine))
+```
+
+The exceptions are effectively syntatic sugar that allows expressing sets of exceptions in a concise way.
+
+### Advantages
```

Comment:
I would like to add a section to talk about the limitation here.  The exception approach only supports `=` operator, if users want to use operators like `pmatch`, `contains`, they still need to user the user_known_xxx macros.

Reply:
Ok, I added it a bit earlier in the discussion of exceptions but I call out equality.